docs/documentation/quartz-3.x/packages/system-text-json.md

```diff
@@ -50,48 +50,62 @@ There's now official solution for migration as there can be quirks in every setu
 **Example hybrid serializer**
 
 ```csharp
-public class MigratorSerializer : IObjectSerializer
+using System.Text.Json;
+
+using Quartz.Simpl;
+using Quartz.Spi;
+
+namespace Quartz;
+
+public sealed class MigratorSerializer : IObjectSerializer
 {
-    private BinaryObjectSerializer binarySerializer;
-    private SystemTextJsonObjectSerializer jsonSerializer;
+    private readonly BinaryObjectSerializer binarySerializer;
+    private readonly SystemTextJsonObjectSerializer jsonSerializer;
 
     public MigratorSerializer()
     {
-        this.binarySerializer = new BinaryObjectSerializer();
+        binarySerializer = new BinaryObjectSerializer();
         // you might need custom configuration, see sections about customizing
         // in documentation
-        this.jsonSerializer = new SystemTextJsonObjectSerializer();
+        jsonSerializer = new SystemTextJsonObjectSerializer();
     }
 
     public T DeSerialize<T>(byte[] data) where T : class
     {
         try
         {
             // Attempt to deserialize data as JSON
-            var result = this.jsonSerializer.DeSerialize<T>(data);
-            return result;
+            return jsonSerializer.DeSerialize<T>(data)!;
         }
-        catch (JsonException
+        catch (JsonException)
         {
             // Presumably, the data was not JSON, we instead use the binary serializer
-            return this.binarySerializer.DeSerialize<T>(data);
+            var binaryData = binarySerializer.DeSerialize<T>(data);
+            if (binaryData is JobDataMap jobDataMap)
+            {
+                // make sure we mark the map as dirty so it will be serialized as JSON next time
+                const string DirtyKey = "____dirty____";
+                jobDataMap[DirtyKey] = "true";
+                jobDataMap.Remove(DirtyKey);
```

Comment:
Yes, internal `isDirty` transforms to true after indexer setter so we can remove the key after and return the dictionary's keys and values to original state, it now only will dirty too with original data inside

Reply:
That's perfect, will give this a go and let you know if it works. 
Thanks a mill Marko.

Reply:
Unfortunately, this didn't work. I've tried your solution but the job stays serialised as binary. I also ran SQL profiler in the background and I could not see any updates to the job details table. I've also tried decorating my job with these 2 attributes but to no avail.

`
	[DisallowConcurrentExecution]
	[PersistJobDataAfterExecution]
	public class QuartzJob : IJob
`

Reply:
As I cannot see what your code does, can you debug against a backup of the database and check how the migration serializer behaves and what is blocking the update. JSON serializer will produce binary output too (UTF-8 bytes) so it won't magically show up as JSON string in the database column.

Reply:
I understand that but the value of the colum shows exactly the same so I don't think it's doing the update, also SQL profiler is not showing any attempt to update that table. I cannot debug the code in Quartz because I cannot load the debug symbols for its code. I'll keep digging.

Reply:
Yes, this is a really hard problem to debug remotely so if you have any means to try out or add more logging etc that would help tremendously.

Reply:
I think this is a bit of a corner case. Current Quartz logic should set job data as `null` when the map is empty, not sure how long this has been so. Can you remember what version of Quartz you started to store data with?

Reply:
The job data map dirty state can indeed get lost. I've tweaked logic a bit further to preserve the information, requires a new release.

Reply:
> I think this is a bit of a corner case. Current Quartz logic should set job data as `null` when the map is empty, not sure how long this has been so. Can you remember what version of Quartz you started to store data with?

Could be as old as Quartz 3.3.2, I can't say for sure, we only migrated to v3.9.0 recently.

Reply:
> The job data map dirty state can indeed get lost. I've tweaked logic a bit further to preserve the information, requires a new release.

Thanks a mill Marko. Any idea when this new release might happen?

Reply:
I saw there is a new version out (3.11) this morning - thank you so much for your help @lahma

Reply:
I tried the fix but it does not seem to work in my case, unfortunately. Every time the job executes (inside the IJob.Execute method) context.JobDetail.JobDataMap.Dirty is false even though it was true after deserialisation. 
If I set it to true manually I get the expected result but it feels wrong to do it there again. It feels like the job data gets initialised again somewhere between deserialisation and execution or else I am doing something wrong/custom in my code. I tried to copy the Quartz source code locally and debug it but no luck. 

public async System.Threading.Tasks.Task Execute(IJobExecutionContext context)
{
    // execute job
    context.JobDetail.JobDataMap[SchedulerConstants.ForceJobDataMapDirty] = "true";
}